maemo-launcher/launcher/invoker.1.in
.\" $Id$
.TH maemo\-invoker 1 2006-02-17 @VERSION@ "Maemo Launcher manual"
.SH NAME
maemo\-invoker \- invoke a program from the launcher
.SH SYNOPSIS
.B maemo\-invoker
.RI [ options "] " file " [" file-options... ]
.SH DESCRIPTION
\fBmaemo\-invoker\fP tells \fBmaemo\-launcher\fP what application to start.
It will invoke the file provided on the command line with its optional
arguments, otherwise it will invoke its own program name with the string
\fI.launch\fP appended and will pass the arguments as well.
So it can be symlinked in place of the real \fIapplication\fP, and this
one moved to \fIapplication.launch\fP.
.SH OPTIONS
These programs follow the usual GNU command line syntax, with long
options starting with two dashes (`-').
A summary of options is included below.
.TP
.B \-\-delay \fISECS\fP
After invoking the program sleep for \fISECS\fP seconds before exiting. If
\fISECS\fP is 0 \fBmaemo\-invoker\fP will wait until the program finishes,
returning its exit status. The default behaviour is to wait for the invoked
program to exit.
.TP
.B \-\-help
Show summary of options.
.TP
.B \-\-version
Show version of program.
.SH ENVIRONMENT
.TP
.B MAEMO_INVOKER_DELAY
Set the exit delay. The command line option overrides this.
.SH SEE ALSO
.BR maemo\-launcher (1).
.br
.SH AUTHOR
\fBmaemo\-invoker\fP was written by Guillem Jover <guillem.jover@nokia.com>.


ChangeLog
2006-02-23  Tommi Komulainen  <tommi.komulainen@nokia.com>

	* gtk/gtkbutton.c (:keyboard-button): Remove obsolete VKB specific
	hack.

2006-02-23  Kristian Rietveld  <kris@imendio.com>

	* gtk/gtktreemodelfilter.c (gtk_tree_model_filter_rows_reordered):
	Make sure we always update the virtual root, also when ->priv->root
	is still NULL.  (Markku Vire)
	N#22651

2006-02-22  Tommi Komulainen  <tommi.komulainen@nokia.com>
	
	* gdk/x11/gdkwindow-x11.c (set_wm_name, gdk_window_set_title,
	gdk_window_set_icon_name): Set non-UTF-8 legacy WM hints only if
	ENABLE_ICCCM_LEGACY is defined. Avoids needlessly loading iconv as it
	wastes quite a bit of memory for no benefit.

2006-02-22  Tommi Komulainen  <tommi.komulainen@nokia.com>

	* */*.c: Add missing Since: tags for custom additions

2006-02-22  Tommi Komulainen  <tommi.komulainen@nokia.com>

	Merge from GNOME CVS. Improves things for GtkFileSystemMemory a bit.

    2006-01-15  Kristian Rietveld  <kris@gtk.org>

	Fixes #324099, Tommi Komulainen.

	* gtk/gtktreestore.[ch] (gtk_tree_store_insert_with_values),
	(gtk_tree_store_insert_with_valuesv): new functions, analog to
	those found in GtkListStore.

	* gtk/gtktreestore.c (gtk_tree_store_real_set_value),
	(gtk_tree_store_set_valist), (gtk_tree_store_set_valist_internal),
	(gtk_tree_store_sort_iter_changed): refactored.

	* gtk/gtk.symbols: updated.

	* tests/treestoretest.c: added a test for this new functionality.

2006-02-17  Tommi Komulainen  <tommi.komulainen@nokia.com>

	Merge from GNOME CVS. Fixes N#22693

    2005-08-27  Matthias Clasen  <mclasen@redhat.com>

	* gdk/x11/gdkproperty-x11.c (gdk_property_get): Don't warn
	when G_MAXLONG is passed as length.

2006-02-17  Tommi Komulainen  <tommi.komulainen@nokia.com>

	Merge from GNOME CVS. Fixes M#352

    2006-01-06  Matthias Clasen  <mclasen@redhat.com>

	* gtk/gtkviewport.c (viewport_set_adjustment): Disconnect from
	the old adjustments signals.  (#325869, Jorn Baayen)

2006-02-16  Tomas Junnonen  <tomas.junnonen@nokia.com>

	* gtk/gtkimcontext.c
	* gtk/gtkimcontext.h
	* gtk/gtkimmulticontext.c: Added "hildon_gtk_im_context_filter_event"
	function
	* gtk/gtktextview.c (gtk_text_view_button_release_event)
	(gtk_text_view_button_press_event): Filter button events with
	imcontext
	* gtk/gtkentry.c (gtk_entry_button_press) (gtk_entry_button_release):
	Filter button events with imcontext

2006-01-31  Tommi Komulainen  <tommi.komulainen@nokia.com>

	* configure.in (LIBTOOL_EXPORT_OPTIONS): Temporarily export a couple
	of private symbols (_gtk_button_set_depressed,
	_gtk_file_chooser_install_properties, _gtk_file_system_create) to work
	around broken hildon.
	* gtk/gtkfilechooserutils.[ch]
	(hildon_gtk_file_chooser_install_properties): Added
	* gtk/gtkbutton.[ch] (hildon_gtk_button_set_depressed): Added

2006-01-31  Mohammad Anwari  <mohammad.anwari@nokia.com>
	
	IM context extension redesigned.

	Added "hildon-input-mode" property and related code:

	* gtk/gtktextview.h
	* gtkentry.h
	* gtk/gtktextview.c (gtk_text_view_class_init)
	(gtk_text_view_set_property) (gtk_text_view_get_property)
	(hildon_gtk_text_view_set_input_mode)
	(hildon_gtk_text_view_get_input_mode) (gtk_text_view_init),
	* gtk/gtkentry.c (gtk_entry_class_init)
	(_gtk_input_mode_convert_old_to_new)
	(_gtk_input_mode_convert_new_to_old) (gtk_entry_set_property)
	(gtk_entry_get_property) (gtk_entry_init)
	(hildon_gtk_entry_set_input_mode) (hildon_gtk_entry_get_input_mode):
	Added "hildon_input_mode" property and deprecated "input_mode" and
	"autocap" properties.
	* gtk/gtkentry.c (hildon_gtk_input_mode_is_valid_char)
	(gtk_entry_filter_text): Rewrote character validation code. Removed
	g_unichar_*() functions that before were used for the validation.
	* gtk/gtkimmulticontext.c (gtk_im_multicontext_set_property)
	(gtk_im_multicontext_get_property) (gtk_im_multicontext_class_init)
	(gtk_im_multicontext_set_slave) (gtk_im_multicontext_notify)
	(gtk_im_multicontext_slave_input_mode_changed_cb): Removed all
	properties from GtkIMMultiContext. Instead keep the
	"hildon-input-mode" property from slave GtkIMContext mirrored.
	* gtk/gtkimcontext.c (gtk_im_context_class_init)
	(gtk_im_context_set_property) (gtk_im_context_get_property)
	(gtk_im_context_init): Added "hildon-input-mode" property.
	* gtk/gtkenums.h: Added HildonGtkInputMode.
	* docs/reference/gtk/tmpl/gtkenums.sgml: Added HildonGtkInputMode
	documentation.

	Added clipboard handling for GtkIMContext:

	* gtk/gtkimcontext.h: Added GtkIMContextClipboardOperation,
	has_selection and clipboard_operation signals,
	hildon_gtk_im_context_has_selection, _copy, _cut and _paste.
	* gtk/gtkimcontext.c (gtk_im_context_class_init): Added
	"has_selection" and "clipboard_operation" signals.
	* gtk/gtkimcontext.c (hildon_gtk_im_context_has_selection)
	(hildon_gtk_im_context_copy) (hildon_gtk_im_context_cut)
	(hildon_gtk_im_context_paste): Added convenience functions for sending
	clipboard handling signals.
	* gtk/gtkimmulticontext.c (gtk_im_multicontext_class_init)
	(gtk_im_multicontext_has_selection_cb)
	(gtk_im_multicontext_clipboard_operation_cb): Forward "has_selection"
	and "clipboard_operation" signals from slave GtkIMContext to ourself.
	* gtk/gtktextview.c (gtk_text_view_init)
	(gtk_text_view_has_selection_handler)
	* gtk/gtkentry.c (gtk_entry_init) (gtk_entry_has_selection_cb)
	(gtk_entry_clipboard_operation_cb): Implemented clipboard
	functionality required by GtkIMContext.

	Misc:

	* gtk/gtktextview.c (gtk_text_view_class_init): Removed unused
	"autocap" style property.
	* gtk/gtkentry.c (gtk_entry_set_visibility): Removed unnecessary code.
	* gtk/gtkimmulticontext.c (get_global_context_id): When fallbacking to
	default locale, g_strdup() returned value so we don't crash.
	* gtk/gtkimcontext.c (gtk_im_context_show) (gtk_im_context_hide):
	Deprecated for hildon_gtk_im_context_show/hide.

	Removed all IM showing/hiding code:

	* gtk/gtktextview.c (gtk_text_view_init)

maemo-launcher/launcher/invoker.c
/*
 * $Id$
 *
 * Copyright (C) 2005, 2006 Nokia
 *
 * Author: Guillem Jover <guillem.jover@nokia.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
 *
 */

#define _GNU_SOURCE

#include <stdio.h>
#include <stdint.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <unistd.h>

#include "config.h"
#include "report.h"
#include "invokelib.h"

#define DEFAULT_DELAY 0

static bool
invoke_recv_ack(int fd)
{
  uint32_t msg;

  /* Revceive ACK. */
  invoke_recv_msg(fd, &msg);

  if (msg != INVOKER_MSG_ACK)
    die(1, "receiving wrong ack (%08x)\n", msg);
  else
    return true;
}

static int
invoker_init(unsigned int delay)
{
  int fd;
  int options = 0;
  struct sockaddr_un sun;

  fd = socket(PF_UNIX, SOCK_STREAM, 0);
  if (fd < 0)
    die(1, "opening invoker socket\n");

  sun.sun_family = AF_UNIX;
  strcpy(sun.sun_path, INVOKER_SOCK);

  if (connect(fd, (struct sockaddr *)&sun, sizeof(sun)) < 0)
    die(1, "connecting to the launcher\n");

  if (!delay)
    options |= INVOKER_MSG_MAGIC_OPTION_WAIT;

  /* Send magic. */
  invoke_send_msg(fd, INVOKER_MSG_MAGIC | INVOKER_MSG_MAGIC_VERSION | options);
  invoke_recv_ack(fd);

  return fd;
}

static bool
invoker_send_exec(int fd, char *exec)
{
  /* Send action. */
  invoke_send_msg(fd, INVOKER_MSG_EXEC);
  invoke_send_str(fd, exec);

  invoke_recv_ack(fd);

  return true;
}

static bool
invoker_send_args(int fd, int argc, char **argv)
{
  int i;

  /* Send action. */
  invoke_send_msg(fd, INVOKER_MSG_ARGS);
  invoke_send_msg(fd, argc);
  for (i = 0; i < argc; i++)
    invoke_send_str(fd, argv[i]);

  invoke_recv_ack(fd);

  return true;
}

static bool
invoker_send_end(int fd)
{
  /* Send action. */
  invoke_send_msg(fd, INVOKER_MSG_END);

  invoke_recv_ack(fd);

  return true;
}

static int
invoker_recv_exit(int fd)
{
  uint32_t msg;

  /* Receive action. */
  invoke_recv_msg(fd, &msg);

  if (msg != INVOKER_MSG_EXIT)
    die(1, "receiving bad exit status (%08x)\n", msg);

  /* Receive status. */
  invoke_recv_msg(fd, &msg);

  return msg;
}

static uint32_t
get_file_int(const char *filename)
{
  FILE *f;
  uint32_t u;

  f = fopen(filename, "r");
  if (f == NULL)
    return 0;

  if (fscanf(f, "%u", &u) < 1)
    u = 0;

  fclose(f);

  return u;
}

/*
 * This function is Linux specific and depends on the lowmem LSM kernel module,
 * but should behave as a nop on systems not supporting it.
 */
static unsigned int
get_linux_lowmem_modifier(void)
{
  uint32_t mem_allowed;
  uint32_t mem_used;
  const unsigned int lowmem_threshold = 80; /* Memory use percentage. */
  int modifier;

  mem_allowed = get_file_int("/proc/sys/vm/lowmem_allowed_pages");
  mem_used = get_file_int("/proc/sys/vm/lowmem_used_pages");

  if (mem_used && mem_allowed)
  {
    modifier = ((mem_used * 100) / mem_allowed) - lowmem_threshold;
    if (modifier < 0)
      modifier = 0;
  }
  else
    modifier = 0;

  return modifier + 1;
}

static unsigned int
get_delay(char *delay_arg)
{
  char *delay_str;
  unsigned int delay;

  if (delay_arg)
    delay_str = delay_arg;
  else
  {
    delay_str = getenv("MAEMO_INVOKER_DELAY");
    /* XXX: Backward compatibility. */
    if (!delay_str)
      delay_str = getenv("MAEMO_INVOKER_TIMEOUT");
  }

  if (delay_str)
    delay = strtoul(delay_str, NULL, 10);
  else
    delay = DEFAULT_DELAY;

  return delay;
}

static void
version(void)
{
  printf("%s (%s) %s\n", PROG_NAME, PACKAGE, VERSION);

  exit(0);
}

static void
usage(int status)
{
  printf("Usage: %s [options] file [file-options]\n"
	 "\n"
	 "Options:\n"
	 "  --delay SECS        After invoking sleep for SECS seconds (default %d).\n"
	 "  --version           Print program version.\n"
	 "  --help              Print this help message.\n",
	 PROG_NAME, DEFAULT_DELAY);

  exit(status);
}

int
main(int argc, char *argv[])
{
  int i;
  int fd;
  int prog_argc = 0;
  char **prog_argv = NULL;
  char *prog_name = NULL;
  int prog_ret = 0;
  char *launch = NULL;
  char *delay_str = NULL;
  unsigned int delay;

  if (strstr(argv[0], PROG_NAME))
  {
    /* Called with our default name. Parse arguments. */
    for (i = 1; i < argc; ++i)
    {
      if (strcmp(argv[i], "--delay") == 0)
      {
	if (argv[++i])
	  delay_str = argv[i];
      }
      else if (strcmp(argv[i], "--version") == 0)
	version();
      else if (strcmp(argv[i], "--help") == 0)
	usage(0);
      else if (strncmp(argv[i], "--", 2) == 0)
	usage(1);
      else
      {
	prog_name = argv[i];
	prog_argc = argc - i;
	prog_argv = &argv[i];
	break;
      }
    }
  }
  else
  {
    /* Called with a different name. Add the proper extension and launch it.
     * Do not try to parse any arguments. */
    asprintf(&launch, "%s.launch", argv[0]);
    prog_name = launch;
    prog_argc = argc;
    prog_argv = argv;

    report_set_output(report_syslog);
  }

  if (!prog_name)
  {
    error("nothing to invoke\n");
    usage(0);
  }

  debug("invoking execution: '%s'\n", prog_name);

  delay = get_delay(delay_str);

  fd = invoker_init(delay);
  invoker_send_exec(fd, prog_name);
  invoker_send_args(fd, prog_argc, prog_argv);
  invoker_send_end(fd);

  if (launch)
    free(launch);

  if (delay)
  {
    /* DBUS cannot cope some times if the invoker exits too early. */
    delay *= get_linux_lowmem_modifier();
    debug("delaying exit for %d seconds\n", delay);
    sleep(delay);
  }
  else
  {
    debug("waiting for invoked program to exit\n");
    prog_ret = invoker_recv_exit(fd);
  }

  close(fd);

  return prog_ret;
}

